src/UI/class-admin-columns-parsely-stats.php

```diff
@@ -0,0 +1,370 @@
+<?php
+/**
+ * UI: Class for adding `Parse.ly Stats` on admin columns
+ *
+ * @package Parsely
+ * @since   3.7.0
+ */
+
+declare(strict_types=1);
+
+namespace Parsely\UI;
+
+use DateTime;
+use Parsely\Parsely;
+use Parsely\RemoteAPI\Remote_API_Base;
+use Parsely\RemoteAPI\Analytics_Posts_API;
+use WP_Screen;
+
+use function Parsely\Utils\get_formatted_number;
+use function Parsely\Utils\get_formatted_time;
+use function Parsely\Utils\get_utc_date_format;
+
+use const Parsely\PARSELY_FILE;
+use const Parsely\Utils\DATE_UTC_FORMAT;
+
+/**
+ * Class for adding `Parse.ly Stats` on admin columns.
+ *
+ * @since 3.7.0
+ *
+ * @phpstan-import-type Analytics_Post_API_Params from Analytics_Posts_API
+ * @phpstan-import-type Analytics_Post from Analytics_Posts_API
+ * @phpstan-import-type Remote_API_Error from Remote_API_Base
+ *
+ * @phpstan-type Parsely_Post_Stats array{
+ *   page_views: string,
+ *   visitors: string,
+ *   avg_time: string,
+ * }
+ *
+ * @phpstan-type Parsely_Posts_Stats_Response array{
+ *   data: array<string, Parsely_Post_Stats>|null,
+ *   error: Remote_API_Error|null,
+ * }
+ */
+class Admin_Columns_Parsely_Stats {
+	/**
+	 * Instance of Parsely class.
+	 *
+	 * @var Parsely
+	 */
+	private $parsely;
+
+	/**
+	 * Internal Variable.
+	 *
+	 * @var WP_Screen|null
+	 */
+	private $current_screen;
+
+	/**
+	 * Published times of visible posts.
+	 *
+	 * Analytics Endpoint don't support post_ids as param so to limit the API we will pass the
+	 * min/max publish time.
+	 *
+	 * @var string[]
+	 */
+	private $utc_published_times = array();
+
+	/**
+	 * Constructor.
+	 *
+	 * @param Parsely $parsely Instance of Parsely class.
+	 */
+	public function __construct( Parsely $parsely ) {
+		$this->parsely = $parsely;
+	}
+
+	/**
+	 * Registers action and filter hook callbacks.
+	 *
+	 * @since 3.7.0
+	 *
+	 * @return void
+	 */
+	public function run(): void {
+		if ( ! $this->parsely->site_id_is_set() || ! $this->parsely->api_secret_is_set() ) {
+			return;
```

Comment:
For `Site ID` there is already a default message as follows

```
The Parse.ly plugin is not active. You need to [provide your Parse.ly Dash Site ID](http://parsely.test/wp-admin/options-general.php?page=parsely) before things get cooking.
```

For `API Secret` there is nothing ...... as far as I understand not all users have API Secret set in the options so if we show error on this then it will show errors for those users everytime.

FYI: If I remove `api_secret` condition then our error parsely stats error message does show `Forbidden` error.

Let me now what approach we should take here. Thanks

Reply:
Does it look broken when the forbidden error is shown?

If not, let's show the error. It is better to signal to the user that something is wrong, and we can also add a note about this in the documentation.

Reply:
No it doesn't looks broken. Is it okay to show this error to all users who are using Parse.ly and doesn't have API Secret because of additional pricing? .... skipped showing error because of this case.

![image](https://user-images.githubusercontent.com/31419912/214259251-fe3d5e40-6847-46bf-9c65-e3c2bcb09fa4.png)

Reply:
Will the error we see in the screenshot be displayed even if the user hides the column?

Reply:
In current implementation .. Yes.

By default this column will appear for our customers so will it be good if they have to hide the column due to this error ..... or we can just keep the current implementation to silently hide it if key isn't set?

Reply:
If this is possible, here is how I would approach this:

- If the column is shown, show the error when no API Secret is set.
- If the column is not shown, do not show the error.

The point is that the error concerns the column, so if the user chooses to hide it, there is no point in displaying the error. However, if the user wants the column, they must be notified that something is wrong. Else they could perceive the lack of data as a bug.

I'd also make the error message more helpful if possible, as we do in the Content Helper sidebar.

Reply:
The recommended approach is possible but is it fine to force all users to disable the column if they are not using the API Secret because by default we are showing the Parse.ly Stats column.

Reply:
If we activate the column by default, we probably do not want to show the error message, because it looks like we're activating a broken feature. @mehmoodak, I guess that this is why you had it hidden.

Reply:
Yes that is the reason.

If we go with `No Default` then it is fine to show the error but it doesn't seems like a good idea to make this feature hidden by default.

Reply:
Reviewing the questions...

`Is it okay to show this error to all users who are using Parse.ly and doesn't have API Secret because of additional pricing?` 
Yes, if we are unable to show Parse.ly data to a VIP customer, we should display an error message that explains the issue and how to resolve it.

Proposed error message:
We are unable to retrieve data for Parse.ly Stats. Please contact support at WordPress VIP for help resolving this issue. 

`Will the error we see in the screenshot be displayed even if the user hides the column?`
Agree that 'if the user chooses to hide it [the column], there is no point in displaying the error. However, if the user wants the column, they must be notified that something is wrong. Else they could perceive the lack of data as a bug.'

`Is it fine to force all users to disable the column if they are not using the API Secret because by default we are showing the Parse.ly Stats column.`
We should avoid forcing users to disable the column. If they don't want to see the column, they have the option to hide it. Otherwise, we want them to know it is available. And, if no data displays in that column, we want to help them find out how to get data to display in the Parse.ly Stats column. 

`There was actually a discussion to hide by default all Content Helper features when an API Secret is not set (this would be a different PR of course). I do not know how much this discussion has progressed and if there was a decision about it.`
There is no decision on this, and if we need to make changes we can do so in a separate PR.

Reply:
> Else they could perceive the lack of data as a bug.

There will be no lack of data because as of now we are hiding the entire column so it should not perceive as bug.

Current Approach is as follows:
- Do not show the Parse.ly Stats column if Site ID and API Secret aren't set.
  - If Site ID isn't set then there is already a message in place which shows the error.
  - If there is no API Secret set then we shows no error and no Parse.ly Stats column i.e. silently hide the column (just in place to avoid any compulsory user action)

Recommended Approach is as follows:
- Do not show the Parse.ly Stats column if Site ID isn't set .... this will show the Site ID error which is already in place.
- For customers that doesn't have API Secret it will show the error and then they have to contact with support or either hide the column.

I think current approach is better because on recommended approach user still have to take some action. Thoughts?

Reply:
I think that the "better approach" boils down to whether we want to show the column by default, even if no Site ID or API Secret is set. So far this is what we have been doing. For example, we currently show the CHS (Content Helper Sidebar) even when no Site ID is set, and even though there is no data to offer (we display a friendly "Contact Us" message instead).

So by auto-hiding the column, we break the previous convention, but we do save the user from the manual step of hiding the column if they do not want it. However, a side-effect of auto-hiding is that many users might totally miss the existence of the column and never know about it (unless they consult the docs or get informed about it in some other way).

This is not really an engineering problem, but rather what's the strategy/approach we use for our CH features. So I'll leave the final decision to @danielabloch about this.

Reply:
As of now the feature doesn't seem broken so for now I will recommend to merge this PR and handle any additional feedback on a separate PR which will make the review simpler.

Apart from this we can also get additional feedback from the team about current implementation.

Reply:
Makes sense to me to move additional feedback on a separate PR. That said, I'll weigh in one more time.

`So by auto-hiding the column, we break the previous convention, but we do save the user from the manual step of hiding the column if they do not want it. However, a side-effect of auto-hiding is that many users might totally miss the existence of the column and never know about it (unless they consult the docs or get informed about it in some other way).`

One of the main objectives with this work is showing people what is possible with Parse.ly data in WordPress. 

- If we auto-hide the column, then people don't know it exists. This goes against our goal to show people what is possible with Parse.ly. 
- We want people to see the column and then decide if they want to use it. In this instance, it is desirable to have people take an action.

Reply:
Column will be visible by default if API Secret is set and as a part of this project I think we will create API Secret for the customers.

With that being said @danielabloch please confirm the final flow. Is it same as you have described above or are there any changes?

Reply:
If API secret is set
- column is visible

If API secret is not set
- column is visible, with messaging on how to activate parsely (contact us)
- this is an edge case for VIP customers
- this is a common case for non-VIP customers

Reply:
Ticket: https://github.com/Parsely/wp-parsely/issues/1330